Add custom buttons to template-card

martindybal opened the pull request:
## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here -->

This PR fixes or closes issue: fixes #
https://github.com/piitaya/lovelace-mushroom/issues/887
https://github.com/piitaya/lovelace-mushroom/issues/1421
## Motivation and Context
As I mentioned in the last issue, I was a bit limited by the template card in some scenarios. It would be really great if I could have a quick action directly on my dashboard.
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested
I tested it locally on my HA instance. I also created a sample in the dev box at `/lovelace-mushroom-showcase/5`.

![image](https://github.com/piitaya/lovelace-mushroom/assets/7644099/2abb6534-63d5-4345-a917-aaa94c7c4702)
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [x] 🚀 New feature (non-breaking change which adds functionality)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] My code follows the code style of this project.
-   [x] My change requires a change to the documentation.
-   [x] I have updated the documentation accordingly.
-   [x] I have tested the change locally.
-   [ ] I followed [the steps](https://github.com/piitaya/lovelace-mushroom#maintainer-steps-to-add-a-new-language) if I add a new language .

BBE-FR commented:
For information, up to what level are the button customisable ?
For example can we like with a template chip:
- define dynamic icon/color ?
-  define tap, hold and double tap actions ?
If yes that is exactly what I wanted to create a room card ! 

Whatever the customization, Many Thanks @martindybal and I hope it goes up to release !

martindybal commented:
@BBE-FR Thank you very much for your feedback and support!

Regarding your questions:

Dynamic icon/color: This is not supported in the current PR, but can be easily added if needed.
Tap, hold and double tap actions: Only tap actions are currently supported, but again, they can be implemented if desired.

Before I proceed with these additions, I need to hear from @piitaya if he wants to finish this PR.

Thanks for your suggestions.

piitaya commented:
Hi 🙂
I like the idea but if we implement this, I think it must be an array instead of 4 elements. 

Also I have concerns to bring new features like this to Mushroom because the gap between Mushroom and tile card become less important with HA updates. Many features request (icon tap action, card features, inverted cover slider, etc...) to Mushroom are improvements done to the tile card so it become hard to do the work in double.

I have some options but I didn't decided about the best solution : 
1. Implement all Mushroom features to tile card and deprecate Mushroom usage
2. Implement all Mushroom features to tile card except template features. Mushroom will become a kind of "advanced tile card" with template support and advanced features.
3. Freeze Mushroom dev and maintain it to avoid regression with HA updates.

The problem with adding this option is that it will encourage people to not use tile card and it can lead to more breaking changes in the future if we want to fully merge Mushroom to Home Assistant.

BBE-FR commented:
@piitaya I imagine that the question is hard...

Option 1 would be Nice...

The lack of template card and chips are the only real things that make me stay with mushroom... without template its difficult to go back to tiles...

For the rest tiles are sligtly less customisable but its enought for 99% of time...

jhemak commented:
+1 for Option 1 ... I would love to switch over to Tile Card 100% but am a heavy user of template and chip.

martindybal commented:
@piitaya I fully understand. The reason for 4 elements is layout, but it's not a problem to use an array with limited size. I'm just not sure about the UI editor. 

I have no problem helping with the development of the Tile Card. I've switched to it myself in a few scenarios with mushrooms. My main reason for sticking with Mushrooms is 
1. Horizontal layout
  - I added card_mod
2. Better tap and hold action management
  - hold_action doesn't work above the icon. It uses icon_tap_action even with hold. 
3. Features like animated icons 

I'm using combination of Mushrooms and tile card for all my room like this
![image](https://github.com/piitaya/lovelace-mushroom/assets/7644099/fc7fe62c-9191-46e4-8af5-4c61b7cbadcf)

BBE-FR commented:
@piitaya 
Do we have any chance to have templates in tile cards the same way as in mushroom ( template defining icon, color, primary and secondary information) ? 

**If not planed** it might be worth to release this feature, I think there is a long lasting request for this kind of function in the mushroom Community. Y think we are numerous to use stack in to fuse template cards (or other mushroom when template is not mandatory) with chips to create additional buttons.
Like in my personal exemple there: https://github.com/piitaya/lovelace-mushroom/issues/55#issuecomment-2046983165

**If planned**, there is an HACS project to allow to use controls: https://github.com/Nerwyn/service-call-tile-feature this could be embeded into the "template tile" to allow to define buttons into this future "_advanced templatable tile_"...

BBE-FR commented:
> @BBE-FR Thank you very much for your feedback and support!
> 
> Regarding your questions:
> 
> Dynamic icon/color: This is not supported in the current PR, but can be easily added if needed. Tap, hold and double tap actions: Only tap actions are currently supported, but again, they can be implemented if desired.
> 
> Before I proceed with these additions, I need to hear from @piitaya if he wants to finish this PR.
> 
> Thanks for your suggestions.

I used chips instead of button with stack in in my cards in order to keep the mushroom look and to have the ability to define the button through graphical interface (including full control of the icons/colors through the template chips)...
Description can be found here: https://github.com/piitaya/lovelace-mushroom/issues/55#issuecomment-2046983165

@martindybal Your proposal is already a great improvement and I fully support this. If at the end you manage to have the same capacity than chips, that would be 100% marvelous to me !

martindybal commented:
@BBE-FR I'm waiting for a reply from @piitaya

martindybal commented:
@piitaya can you please comment? Does it make sense to continue on this or would TileCard be the future?

piitaya commented:
Sorry for the late reply, I was busy these days. 
I'm not sure tile card will have templating features because it's too complex for most users and it can lead to performance issue if there are many templates.
However, I do agree that Mushroom should become an "advanced tile card" with many features, And of course, these features should be edited with a GUI.

For now, I would say we close this PR and I will work on a "advanced" version of the card but I have no deadline to share... However, I personally only use Mushroom card for the template card so I would love to move forward 😅

BBE-FR commented:
> I will work on a "advanced" version of the card but I have no deadline to share... However, I personally only use Mushroom card for the template card so I would love to move forward

This might be a way to advance on the subject:
instead of updating endlessly mushrooms or risking to have to much breaking changes as stated here: https://github.com/piitaya/lovelace-mushroom/issues/1259#issuecomment-2325865362

One option could be:
- to create a new "advanced tile" custom card in HACS (possibly called called toadstool ??) that would basically be a tile card structure and look, but with full control on icon and templating. 
- to let the mushroom as it is, without new functionalities, and only maintain it, as long as feasible.
- This card (or card family) could then, evolve to implement new functionalities (like buttons, sliders, etc, with same design as the tile card, but fully controllable by the user).

I think that otherwise, there will always be pros and cons to stay with mushroom "classical" cards and some will never migrate to tiles.

My personal example is as follow: 
I am willing to go full Tiles (despite I prefer the interface of mushroom with collapsible controls and selectors) but restrained by the lack of:
- template card (more than 75% of my mushroom cards)
- chips (that can be placed anywhere and not just "on top")

With the latest badges (especially if they can be placed on top of sections) I could get rid of chips and use only Tiles. The last real functional blocking point for me is templating capability in Tile or Tile-like card...

j9brown commented:
I'm debating cherry-picking this PR into my own branch of Mushroom because it solves an immediate problem I have with wanting to add buttons to end of template cards with a horizontal appearance.  Tile doesn't support a horizontal appearance currently, so I'm pretty much stuck using Mushroom.

But if this is the way the wind is blowing, then maybe I should redo my layouts to not rely on the horizontal appearance.

To be honest, I think that it would be better to expand the built-in Tile card's feature set to cover the needs of power users too rather than to make a separate "advanced Tile" card that's sort of the same and sort of different and sort of awkward to find and with significant duplication of development effort overall.  Expanding the built-in Tile card (or making more flavors of built-in Tile cards) would help users progressively discover the capabilities they need as they need them.  It also helps the Home Assistant user base as a whole converge on more of the same design vocabulary.  Some programs address this progressive discovery issue by hiding more complex or less frequently used features behind toggles such that a hard bifurcation of the UI isn't necessary.

martindybal commented:
@j9brown I do exactly that. I moved whole dashboard to TileCard with service call tile customer feature. I’m using card_mod for horizontal layout. 
![image](https://github.com/user-attachments/assets/8757cc1a-3700-4664-b586-946f2aaf8645)

piitaya commented:
Thank you for the contribution. 
Unfortunately, I'm closing this PR as I'm planning to release a similar feature in the following months with integration with tile card and card features.
